tests/test_experiments.py

```diff
@@ -0,0 +1,123 @@
+
+    log_dir = test_dir / 'log'
+    if os.path.exists(log_dir):
+        shutil.rmtree(log_dir)
+
+    model_data_dir = test_dir / 'test_model_data'
+    if os.path.exists(model_data_dir):
+        shutil.rmtree(model_data_dir)
+
+
+def get_test_dir() -> pathlib.Path:
+    """Return the path to the directory containing these tests."""
+    return pathlib.Path(__file__).parent
+
+
+def run_python_script(script: str | pathlib.Path, script_args: list[str]) -> str:
+    """
+    Run the provided Python script and return its output.
+    Use the same Python interpreter as the current process.
+    """
+
+    def run_shell_command(command: str, cmd_args: list[str]) -> str:
+        """Run the provided shell command in the specified working directory and return its output."""
+        output = subprocess.run(
+            [command] + cmd_args,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.STDOUT,
+            cwd=get_test_dir(),
+        )
+
+        if output.returncode != 0:
+            raise RuntimeError(
+                f"Command {command} with args {cmd_args} failed with error code {output.returncode}:\n{output.stderr}"
+            )
+
+        return output.stdout.decode()
+
+    python = str(sys.executable)
+    py_args = [str(script)] + script_args
+    return run_shell_command(python, py_args)
+
+
+def check_ppl_in_log(log: str, expected_ppl: float | None, expected_parameters: int | None) -> None:
+    """Verify that the log contains the expected perplexity and parameters result."""
+
+    def verify(pattern: str, value: float | int) -> None:
+        match = re.search(pattern, log)
+        assert match, f'Expected to find pattern {pattern} in the log'
+        assert np.isclose(
+            float(match.group(1).replace(',', '')), value, atol=5e-1, rtol=1e-2
+        ), f'Expected {value} but got {match.group(1)}'
+
+    if expected_ppl is not None:
+        verify(r'(?:After rotating and slicing|Loaded model perplexity:) (\d+\.\d+)', expected_ppl)
+
+    if expected_parameters is not None:
+        verify(r'Sliced model parameters: ([0-9,]+)', expected_parameters)
+
+
+def verify_run_slicegpt_perplexity(model: str, sparsity: float, expected_ppl: float, expected_parameters: int) -> None:
+    """Test the run_slicegpt_perplexity.py script with the provided parameters."""
+    # test rotate, slice and save model
+    tests_dir = get_test_dir()
+    script = tests_dir.parent / 'experiments' / 'run_slicegpt_perplexity.py'
+    save_dir = tests_dir / 'test_model_data'
+    args = ['--no-wandb', '--model', str(model)]
+
+    ext_args = ['--sparsity', str(sparsity), '--save-dir', str(save_dir)]
+    log = run_python_script(script, args + ext_args)
+    check_ppl_in_log(log, expected_ppl=expected_ppl, expected_parameters=expected_parameters)
+
+    # test load and slice model
+    model_path = save_dir / (model.split('/')[-1] + '_' + str(sparsity) + '.pt')
+    ext_args = ['--sliced-model-path', str(model_path)]
```

Comment:
Do we not need to pass `--sparsity` anymore when loading a model? It's possible I missed this in an earlier PR.

Reply:
No, we don't, and so it's not passed to the actual target script. We use it here to determine which checkpoint to use. So a few lines above we sliced and saved a model with sparsity 0.2, therefore we look for that checkpoint here (...0_2.pt), load it and confirm it gives the same ppl.